### ISSUE TITLE
Memory cannot be released when receiving files in form-data

### ISSUE THREAD

**Issue by 20164930** [zeromicro/go-zero#3564]
**Describe the bug**

I created a handler to receive files uploaded by clients. In the handler, I used the ParseMultipartForm function to set the maximum memory, and used FormFile to receive files and write them locally. However, during the debugging process, I found that the memory occupied by the Buffer could not be released. The strange thing is that this doesn't happen when I create an http server using net/http, I'm not sure if this problem is related to go-zero.

**To Reproduce**

1. The code is
```go
  func main() {
	go func() {
		http.ListenAndServe("0.0.0.0:8000", nil)
	}()
	
	var restConf rest.RestConf
	conf.MustLoad("src/init.yaml", &restConf)
	s, err := rest.NewServer(restConf)
	if err != nil {
		log.Fatal(err)
		return
	}

	s.AddRoute(rest.Route{
		Method:  http.MethodPost,
		Path:    "/upload",
		Handler: upload,
	})
	defer s.Stop()
	s.Start()

	//http.HandleFunc("/upload", upload)
	//http.ListenAndServe("0.0.0.0:8888", nil)
}

func upload(w http.ResponseWriter, request *http.Request) {
	if err := request.ParseMultipartForm(32 << 20); err != nil {
		...
	}
	defer request.MultipartForm.RemoveAll()
	file, header, err := request.FormFile("file")
	if err != nil {
		...
	}
	defer file.Close()
	dst, err := os.Create(header.Filename)
	if err != nil {
		...
	}
	defer dst.Close()

	_, err = io.Copy(dst, file)
	if err != nil {
		...
	}
}
   
   ```

2. The error is

The memory allocation analyzed by pprof is as follows:
![image](https://github.com/zeromicro/go-zero/assets/43667525/ca1cdf1a-c219-422f-adbe-3976a01368c5)

Memory allocation when using net/http is as follows:
![image](https://github.com/zeromicro/go-zero/assets/43667525/3675e8be-f903-45fb-ab5e-0be144b8852b)

**Expected behavior**
Please confirm if this problem is related to go-zero, and how can I receive a large file using less memory?

**Screenshots**

**Environments (please complete the following information):**
 - OS: windows10 & centos7
 - go-zero version: 1.5.5

**More description**

**Comment by 20164930** [comment]
The iox.DupReadCloser() function is called in the LogHandler() function in the rest/handler/loghandler.go file,The content of this function is:
```
// DupReadCloser returns two io.ReadCloser that read from the first will be written to the second.
// The first returned reader needs to be read first, because the content
// read from it will be written to the underlying buffer of the second reader.
func DupReadCloser(reader io.ReadCloser) (io.ReadCloser, io.ReadCloser) {
	var buf bytes.Buffer
	tee := io.TeeReader(reader, &buf)
	return ioutil.NopCloser(tee), ioutil.NopCloser(&buf)
}
```
This function consumes a lot of memory.

**Comment by 20164930** [comment]
I'm sure I'm checking the memory in use：
```
$ go tool pprof --inuse_space http://127.0.0.1:8888/debug/pprof/heap
Fetching profile over HTTP from http://127.0.0.1:8888/debug/pprof/heap
Saved profile in C:\Users\User\pprof\pprof.___1go_build_main_go.exe.alloc_objects.alloc_space.inuse_objects.inuse_space.005.pb.gz
File: ___1go_build_main_go.exe
Build ID: C:\Users\User\AppData\Local\Temp\GoLand\___1go_build_main_go.exe2023-11-07 10:05:01.6110518 +0800 CST
Type: inuse_space
Time: Nov 7, 2023 at 10:13am (CST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 388MB, 99.87% of 388.50MB total
Dropped 10 nodes (cum <= 1.94MB)
Showing top 10 nodes out of 24
      flat  flat%   sum%        cum   cum%
     388MB 99.87% 99.87%      388MB 99.87%  bytes.growSlice
         0     0% 99.87%      384MB 98.84%  bufio.(*Reader).Peek
         0     0% 99.87%      384MB 98.84%  bufio.(*Reader).fill
         0     0% 99.87%        4MB  1.03%  bytes.(*Buffer).ReadFrom
         0     0% 99.87%      384MB 98.84%  bytes.(*Buffer).Write
         0     0% 99.87%      388MB 99.87%  bytes.(*Buffer).grow
         0     0% 99.87%      388MB 99.87%  github.com/zeromicro/go-zero/rest/handler.(*timeoutHandler).ServeHTTP.func1
         0     0% 99.87%      388MB 99.87%  github.com/zeromicro/go-zero/rest/handler.GunzipHandler.func1
         0     0% 99.87%      388MB 99.87%  github.com/zeromicro/go-zero/rest/handler.MaxBytesHandler.func2.1
         0     0% 99.87%      388MB 99.87%  github.com/zeromicro/go-zero/rest/handler.MetricHandler.func1.1

```
traces:

```
(pprof)  traces bytes.growSlice
File: ___1go_build_main_go.exe
Build ID: C:\Users\User\AppData\Local\Temp\GoLand\___1go_build_main_go.exe2023-11-07 10:05:01.6110518 +0800 CST
Type: inuse_space
Time: Nov 7, 2023 at 10:13am (CST)
-----------+-------------------------------------------------------
     bytes:  512MB
         0   bytes.growSlice
             bytes.(*Buffer).grow
             bytes.(*Buffer).Write
             io.(*teeReader).Read
             mime/multipart.(*stickyErrorReader).Read
             bufio.(*Reader).fill
             bufio.(*Reader).Peek
             mime/multipart.partReader.Read
             mime/multipart.(*Part).Read
             io.copyBuffer
             io.CopyBuffer
             mime/multipart.(*Reader).readForm
             mime/multipart.(*Reader).ReadForm (inline)
             net/http.(*Request).ParseMultipartForm
             main.upload
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.GunzipHandler.func1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.MaxBytesHandler.func2.1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.MetricHandler.func1.1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.RecoverHandler.func1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.(*timeoutHandler).ServeHTTP.func1
-----------+-------------------------------------------------------
     bytes:  256MB
     256MB   bytes.growSlice
             bytes.(*Buffer).grow
             bytes.(*Buffer).Write
             io.(*teeReader).Read
             mime/multipart.(*stickyErrorReader).Read
             bufio.(*Reader).fill
             bufio.(*Reader).Peek
             mime/multipart.partReader.Read
             mime/multipart.(*Part).Read
             io.copyBuffer
             io.CopyBuffer
             mime/multipart.(*Reader).readForm
             mime/multipart.(*Reader).ReadForm (inline)
             net/http.(*Request).ParseMultipartForm
             main.upload
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.GunzipHandler.func1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.MaxBytesHandler.func2.1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.MetricHandler.func1.1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.RecoverHandler.func1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.(*timeoutHandler).ServeHTTP.func1
-----------+-------------------------------------------------------
     bytes:  128MB
     128MB   bytes.growSlice
             bytes.(*Buffer).grow
             bytes.(*Buffer).Write
             io.(*teeReader).Read
             mime/multipart.(*stickyErrorReader).Read
             bufio.(*Reader).fill
             bufio.(*Reader).Peek
             mime/multipart.partReader.Read
             mime/multipart.(*Part).Read
             io.copyBuffer
             io.CopyBuffer
             mime/multipart.(*Reader).readForm
             mime/multipart.(*Reader).ReadForm (inline)
             net/http.(*Request).ParseMultipartForm
             main.upload
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.GunzipHandler.func1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.MaxBytesHandler.func2.1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.MetricHandler.func1.1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/rest/handler.RecoverHandler.func1
             net/http.HandlerFunc.ServeHTTP
             github.com/zeromicro/go-zero/re
```
According to my testing, there are actually two problems:
1. The memory used during the upload process should not exceed 32<<20, but it actually exceeds
2. After the upload is completed, it will take a few minutes for the memory to be recycled instead of being released immediately.

I suggest you observe it during the upload process or immediately after the upload is completed. In addition to pprof, you can also use the Windows Resource Monitor or the Linux command line to observe the memory changes of the process.

**Comment by haobinc** [comment]
i meet this problem too!

**Comment by Luo-Yihang** [comment]
@kevwan I can reproduce the issue. Specifically, my  investigation revealed that the 'ParseMultipartForm' function in go-zero is causing memory leaks and even unresolved temporary file issues when the actual file size exceeds the 'maxMemory' parameter set in 'ParseMultipartForm'. This issue can be validated through [go-zero official example](https://github.com/zeromicro/zero-examples/tree/main/http/uploadhttp://example.com) (Please try to upload a large files, e.g. 2gb to better reproduce the bug), where I was able to reproduce the memory leak when utilizing 'ParseMultipartForm'. Additionally, I conducted tests on an isolated Go file, which demonstrated normal functionality without encountering any issues.

**Comment by liang09255** [comment]
+1

**Comment by kevwan** [comment]
After re-investigating into this issue, I still didn't find any memory leakage.

Remember to force GC before digging into the profiling results. For example, I'm using command like `go tool pprof --inuse_space "http://localhost:6060/debug/pprof/heap?gc=1"`, after uploading couple gigabytes large files.

<img width="1003" height="358" alt="Image" src="https://github.com/user-attachments/assets/14d42577-a7e2-4b0f-8cf8-42fd09705538" />

Also, I used the builtin continuous profiling of go-zero, looks like below:

<img width="1422" height="304" alt="Image" src="https://github.com/user-attachments/assets/9fdc81fd-4007-4995-a68b-ba6fe7271da4" />

**Comment by 20164930** [comment]
ParseMultipartForm() can temporarily store data on disk when parsing form-data, so as to reduce memory usage. However, this method does not work in go-zero, and all data will be loaded into memory. Maybe the problem lies here.

**Comment by kevwan** [comment]
The memory will be released after the request is done.

**Comment by 20164930** [comment]
In my test, the memory is not released immediately after calling MultipartForm.RemoveAll(), but is automatically released after a period of time after the request ends. So I guess a component of go-zero also caches a copy of the form-data data, and puts all the body data into memory, instead of caching part of the data on disk like ParseMultipartForm(). This causes a lot of memory to be used when uploading a large file.

**Comment by liang09255** [comment]
> The memory will be released after the request is done.

I don't think this heap memory occupation is reasonable. In streaming writing files, the memory occupied should be the size of buffer. If the heap memory equal to the file size is occupied instantaneously, it will lead to that large files cannot be uploaded due to the system memory size, and uploading files exceeding the memory size will lead to service OOM.

**Comment by liang09255** [comment]
> In my test, the memory is not released immediately after calling MultipartForm.RemoveAll(), but is automatically released after a period of time after the request ends. So I guess a component of go-zero also caches a copy of the form-data data, and puts all the body data into memory, instead of caching part of the data on disk like ParseMultipartForm(). This causes a lot of memory to be used when uploading a large file.

Did you finally find any solution to this problem?

**Comment by kevwan** [comment]
> > The memory will be released after the request is done.
> 
> I don't think this heap memory occupation is reasonable. In streaming writing files, the memory occupied should be the size of buffer. If the heap memory equal to the file size is occupied instantaneously, it will lead to that large files cannot be uploaded due to the system memory size, and uploading files exceeding the memory size will lead to service OOM.

I agree with you. I'll try to figure out a solution to avoid the usage of `iox.DupReadCloser()` if the content length is a large value.

**Comment by liang09255** [comment]
> > > The memory will be released after the request is done.
> > 
> > 
> > I don't think this heap memory occupation is reasonable. In streaming writing files, the memory occupied should be the size of buffer. If the heap memory equal to the file size is occupied instantaneously, it will lead to that large files cannot be uploaded due to the system memory size, and uploading files exceeding the memory size will lead to service OOM.
> 
> I agree with you. I'll try to figure out a solution to avoid the usage of `iox.DupReadCloser()` if the content length is a large value.

This problem may not be caused by go-zero, because I tried to use `MultipartReader` to manually process the file reading logic, and this problem still appeared.

```
(pprof) traces grow
File: xxx
Type: inuse_space
Time: Jul 26, 2025 at 11:07pm (CST)
-----------+-------------------------------------------------------
     bytes:  256MB
     256MB   bytes.growSlice
             bytes.(*Buffer).grow
             bytes.(*Buffer).Write
             io.(*teeReader).Read
             net/http.(*maxBytesReader).Read
             mime/multipart.(*stickyErrorReader).Read
             bufio.(*Reader).fill
             bufio.(*Reader).Peek
             mime/multipart.partReader.Read
             mime/multipart.(*Part).Read (inline)
             myproject/utils.myfunc.func2
-----------+-------------------------------------------------------
     bytes:  128MB
         0   bytes.growSlice
             bytes.(*Buffer).grow
             bytes.(*Buffer).Write
             io.(*teeReader).Read
             net/http.(*maxBytesReader).Read
             mime/multipart.(*stickyErrorReader).Read
             bufio.(*Reader).fill
             bufio.(*Reader).Peek
             mime/multipart.partReader.Read
             mime/multipart.(*Part).Read (inline)
             myproject/utils.myfunc.func2
```

**Comment by kevwan** [comment]
> > > > The memory will be released after the request is done.
> > > 
> > > 
> > > I don't think this heap memory occupation is reasonable. In streaming writing files, the memory occupied should be the size of buffer. If the heap memory equal to the file size is occupied instantaneously, it will lead to that large files cannot be uploaded due to the system memory size, and uploading files exceeding the memory size will lead to service OOM.
> > 
> > 
> > I agree with you. I'll try to figure out a solution to avoid the usage of `iox.DupReadCloser()` if the content length is a large value.
> 
> This problem may not be caused by go-zero, because I tried to use `MultipartReader` to manually process the file reading logic, and this problem still appeared.
> 
> ```
> (pprof) traces grow
> File: xxx
> Type: inuse_space
> Time: Jul 26, 2025 at 11:07pm (CST)
> -----------+-------------------------------------------------------
>      bytes:  256MB
>      256MB   bytes.growSlice
>              bytes.(*Buffer).grow
>              bytes.(*Buffer).Write
>              io.(*teeReader).Read
>              net/http.(*maxBytesReader).Read
>              mime/multipart.(*stickyErrorReader).Read
>              bufio.(*Reader).fill
>              bufio.(*Reader).Peek
>              mime/multipart.partReader.Read
>              mime/multipart.(*Part).Read (inline)
>              myproject/utils.myfunc.func2
> -----------+-------------------------------------------------------
>      bytes:  128MB
>          0   bytes.growSlice
>              bytes.(*Buffer).grow
>              bytes.(*Buffer).Write
>              io.(*teeReader).Read
>              net/http.(*maxBytesReader).Read
>              mime/multipart.(*stickyErrorReader).Read
>              bufio.(*Reader).fill
>              bufio.(*Reader).Peek
>              mime/multipart.partReader.Read
>              mime/multipart.(*Part).Read (inline)
>              myproject/utils.myfunc.func2
> ```

Did you use `?gc=1`?

**Comment by liang09255** [comment]
The root cause of this issue is that the logging middleware `DetailedLogHandler` uses `teeReader` to obtain an additional complete copy of the body for logging purposes, resulting in the entire large file being read into memory.  

This issue can be triggered when both `Verbose` and `Log` are set to `true`.  

I suggest adding a maximum body log length limit, such as 4KB, when extracting the body content to prevent this issue from occurring. In most cases, excessively large body logs are not helpful for debugging.  

If you think this approach is feasible, I can try to submit a PR to fix this issue.

**Comment by kevwan** [comment]
> The root cause of this issue is that the logging middleware `DetailedLogHandler` uses `teeReader` to obtain an additional complete copy of the body for logging purposes, resulting in the entire large file being read into memory.
> 
> This issue can be triggered when both `Verbose` and `Log` are set to `true`.
> 
> I suggest adding a maximum body log length limit, such as 4KB, when extracting the body content to prevent this issue from occurring. In most cases, excessively large body logs are not helpful for debugging.
> 
> If you think this approach is feasible, I can try to submit a PR to fix this issue.

Thanks for your help! Would you please submit a PR to fix it?

**Comment by liang09255** [comment]
> > The root cause of this issue is that the logging middleware `DetailedLogHandler` uses `teeReader` to obtain an additional complete copy of the body for logging purposes, resulting in the entire large file being read into memory.
> > This issue can be triggered when both `Verbose` and `Log` are set to `true`.
> > I suggest adding a maximum body log length limit, such as 4KB, when extracting the body content to prevent this issue from occurring. In most cases, excessively large body logs are not helpful for debugging.
> > If you think this approach is feasible, I can try to submit a PR to fix this issue.
> 
> Thanks for your help! Would you please submit a PR to fix it?

sure

**Comment by kevwan** [comment]
I've submitted a PR to fix this. Please kindly review the changes.
